### PR TITLE
[MAINTENANCE] Clean Up Type Hints and Minor Refactoring For Better Code Elegance/Readability

### PR DESCRIPTION
**Pull request by alexsherstinsky**
### Scope
* Fix incorrect capitalization of `Dict` type hint that does not specify the key and value types.
* Fix type hint warning in `validator.py` by adding `Union` of proper type for Batch list.
* Split `MetricsCalculator.compute_metrics()` into `ValidationGraph` construction phase and graph resolution phase.
* Add descriptive docstrings to metric computation methods.
* Minor formatting clean up for better code readability.

Please annotate your PR title to describe what the PR does, then give a brief bulleted description of your PR below. PR titles should begin with [BUGFIX], [FEATURE],  [DOCS], or [MAINTENANCE]. If a new feature introduces breaking changes for the Great Expectations API or configuration files, please also add [BREAKING]. You can read about the tags in our [contributor checklist](https://docs.greatexpectations.io/docs/contributing/contributing_checklist).

Changes proposed in this pull request:
- JIRA: GREAT-1347/GREAT-181/GREAT-1413
-
-

After submitting your PR, CI checks will run and @cla-bot will check for your CLA signature.

For a PR with nontrivial changes, we review with both design-centric and code-centric lenses.

In a design review, we aim to ensure that the PR is consistent with our relationship to the open source community, with our software architecture and abstractions, and with our users' needs and expectations. That review often starts well before a PR, for example in github issues or slack, so please link to relevant conversations in notes below to help reviewers understand and approve your PR more quickly (e.g. `closes #123`).

Previous Design Review notes:
-
-
-

### Definition of Done
Please delete options that are not relevant.

- [x] My code follows the Great Expectations [style guide](https://docs.greatexpectations.io/docs/contributing/style_guides/code_style)
- [x] I have performed a [self-review](https://docs.greatexpectations.io/docs/contributing/contributing_checklist) of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added [unit tests](https://docs.greatexpectations.io/docs/contributing/contributing_test#writing-unit-and-integration-tests) where applicable and made sure that new and existing tests are passing.
- [x] I have run any local integration tests and made sure that nothing is broken.

Thank you for submitting!

### CODE REVIEW

File: great_expectations/execution_engine/execution_engine.py [hunk 1]
```diff
@@ -592,7 +592,7 @@ class MetricDomainTypes.
 
     @staticmethod
     def _split_table_metric_domain_kwargs(
-        domain_kwargs: Dict,
+        domain_kwargs: dict,
```
Comment (cdkini) [inline great_expectations/execution_engine/execution_engine.py:595]: I don't mind either way but do we have a convention for `dict` vs `Dict`? Don't think it's worth splitting hairs over (as long as they both mean `Dict[str, Any]`)

File: great_expectations/execution_engine/sqlalchemy_execution_engine.py [hunk 2]
```diff
@@ -544,7 +544,7 @@ def _get_sqlalchemy_key_pair_auth_url(
 
     def get_domain_records(  # noqa: C901 - 24
         self,
-        domain_kwargs: Dict,
+        domain_kwargs: dict,
```
Comment (cdkini) [inline great_expectations/execution_engine/sqlalchemy_execution_engine.py:547]: In the case we remove all `Dict`, please remove the import as well

File: great_expectations/validator/metrics_calculator.py [hunk 3]
```diff
@@ -140,32 +140,28 @@ def get_metrics(
     def compute_metrics(
         self,
         metric_configurations: List[MetricConfiguration],
+        runtime_configuration: Optional[dict] = None,
```
Comment (cdkini) [inline great_expectations/validator/metrics_calculator.py:143]: I'm always wary of updating public signatures - why is this necessary?

File: great_expectations/execution_engine/sqlalchemy_execution_engine.py (hunk 2 above)
Comment (alexsherstinsky) [reply great_expectations/execution_engine/sqlalchemy_execution_engine.py:547]: @cdkini Thanks!  I think that the linter should tell me as much.  Checking now.

File: great_expectations/validator/metrics_calculator.py (hunk 3 above)
Comment (alexsherstinsky) [reply great_expectations/validator/metrics_calculator.py:143]: @cdkini Because it was an omission before -- the methods it calls require this, and we should uphold it.  The optional aspect should make any public usage seamless.  Thanks.

File: great_expectations/execution_engine/execution_engine.py (hunk 1 above)
Comment (alexsherstinsky) [reply great_expectations/execution_engine/execution_engine.py:595]: @cdkini I think that `Dict` without the types for key and value is wrong.

File: great_expectations/execution_engine/sqlalchemy_execution_engine.py (hunk 2 above)
Comment (alexsherstinsky) [reply great_expectations/execution_engine/sqlalchemy_execution_engine.py:547]: @cdkini In this case, it seems that this import is used elsewhere.

File: great_expectations/validator/metrics_calculator.py [hunk 4]
```diff
@@ -176,3 +172,32 @@ def compute_metrics(
...
+        Returns:
+            Resulting "ValidationGraph" object.
+        """
+        graph: ValidationGraph = ValidationGraph(
+            execution_engine=self._execution_engine
+        )
+
+        metric_configuration: MetricConfiguration
+        for metric_configuration in metric_configurations:
+            graph.build_metric_dependency_graph(
```
Comment (NathanFarmer) [inline great_expectations/validator/metrics_calculator.py:198]: ❤️ the recursive solution

**Review by NathanFarmer — APPROVED** [review]
